Bad selector format causes SIGSEGV 0.33.5

Back for more...

Passing a selector without proper k=v format causes SIGSEGV

`helmfile -e stage -l thisisgonnabreak status`

```
2018/09/09 00:09:16 Malformed label: thisisgonnabreak. Expected label in form k=v or k!=v
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0x766eb4]

goroutine 1 [running]:
github.com/roboll/helmfile/vendor/github.com/urfave/cli.HandleAction.func1(0xc420184f48)
	/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/app.go:474 +0x279
panic(0x7d10a0, 0xa79ae0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
main.(*app).FindAndIterateOverDesiredStates(0xc420184a80, 0x0, 0x0, 0xc420184af0, 0x0, 0x0, 0xc4200af0b0, 0x1, 0x1, 0x7ffdef97e080, ...)
	/go/src/github.com/roboll/helmfile/main.go:630 +0x524
main.findAndIterateOverDesiredStatesUsingFlags(0xc420100c80, 0xc420184af0, 0xc420184b10, 0x16)
	/go/src/github.com/roboll/helmfile/main.go:568 +0x2f3
main.main.func8(0xc420100c80, 0x0, 0x0)
	/go/src/github.com/roboll/helmfile/main.go:390 +0x46
reflect.Value.call(0x7b7380, 0x859ab8, 0x13, 0x83a5cf, 0x4, 0xc420184ee8, 0x1, 0x1, 0xc4200b0040, 0x82de60, ...)
	/usr/local/go/src/reflect/value.go:447 +0x969
reflect.Value.Call(0x7b7380, 0x859ab8, 0x13, 0xc420184ee8, 0x1, 0x1, 0x4, 0xc4200cd348, 0x5)
	/usr/local/go/src/reflect/value.go:308 +0xa4
github.com/roboll/helmfile/vendor/github.com/urfave/cli.HandleAction(0x7b7380, 0x859ab8, 0xc420100c80, 0x0, 0x0)
	/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/app.go:483 +0x173
github.com/roboll/helmfile/vendor/github.com/urfave/cli.Command.Run(0x83c546, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x84e2e1, 0x29, 0x0, ...)
	/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/command.go:186 +0x999
github.com/roboll/helmfile/vendor/github.com/urfave/cli.(*App).Run(0xc4200aa300, 0xc4200ac000, 0x6, 0x6, 0x0, 0x0)
	/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/app.go:237 +0x546
main.main()
	/go/src/github.com/roboll/helmfile/main.go:474 +0x1f79
```

--- a/app_test.go
+++ b/app_test.go
@@ -198,3 +198,107 @@ releases:
 		}
 	}
 }
+
+// See https://github.com/roboll/helmfile/issues/322
+func TestFindAndIterateOverDesiredStates_Selectors(t *testing.T) {
+	absPaths := map[string]string{
+		".": "/path/to",
+		"/path/to/helmfile.d": "/path/to/helmfile.d",
+	}
+	dirs := map[string]bool{
+		"helmfile.d": true,
+	}
+	files := map[string]string{
+		"helmfile.yaml": `
+helmfiles:
+- helmfile.d/a*.yaml
+- helmfile.d/b*.yaml
+`,
+		"/path/to/helmfile.d/a1.yaml": `
+releases:
+- name: zipkin
+  chart: stable/zipkin
+`,
+		"/path/to/helmfile.d/a2.yaml": `
+releases:
+- name: prometheus
+  chart: stable/prometheus
+`,
+		"/path/to/helmfile.d/b.yaml": `
+releases:
+- name: grafana
+  chart: stable/grafana
+`,
+	}
+	globMatches := map[string][]string{
+		"/path/to/helmfile.d/a*.yaml": []string{"/path/to/helmfile.d/a1.yaml", "/path/to/helmfile.d/a2.yaml"},
+		"/path/to/helmfile.d/b*.yaml": []string{"/path/to/helmfile.d/b.yaml"},
+	}
+	fileExistsAt := func(path string) bool {
+		_, ok := files[path]
+		return ok
+	}
+	directoryExistsAt := func(path string) bool {
+		_, ok := dirs[path]
+		return ok
+	}
+	readFile := func(filename string) ([]byte, error) {
+		str, ok := files[filename]
+		if !ok {
+			return []byte(nil), fmt.Errorf("no file found: %s", filename)
+		}
+		return []byte(str), nil
+	}
+	glob := func(pattern string) ([]string, error) {
+		matches, ok := globMatches[pattern]
+		if !ok {
+			return []string(nil), fmt.Errorf("no file matched: %s", pattern)
+		}
+		return matches, nil
+	}
+	abs := func(path string) (string, error) {
+		a, ok := absPaths[path]
+		if !ok {
+			return "", fmt.Errorf("abs: unexpected path: %s", path)
+		}
+		return a, nil
+	}
+	app := &app{
+		readFile:          readFile,
+		glob:              glob,
+		abs:               abs,
+		fileExistsAt:      fileExistsAt,
+		directoryExistsAt: directoryExistsAt,
+		kubeContext:       "default",
+		logger:            helmexec.NewLogger(os.Stderr, "debug"),
+	}
+	noop := func(st *state.HelmState, helm helmexec.Interface) []error {
+		return []error{}
+	}
+
+	testcases := []struct {
+		label     string
+		expectErr bool
+		errMsg    string
+	}{
+		{label: "name=prometheus", expectErr: false},
+		{label: "name=", expectErr: true, errMsg: "Malformed label: name=. Expected label in form k=v or k!=v"},
+		{label: "name!=", expectErr: true, errMsg: "Malformed label: name!=. Expected label in form k=v or k!=v"},
+		{label: "name", expectErr: true, errMsg: "Malformed label: name. Expected label in form k=v or k!=v"},
+	}
+
+	for _, testcase := range testcases {
+		err := app.FindAndIterateOverDesiredStates(
+			"helmfile.yaml", noop, "", []string{testcase.label}, "default",
+		)
+		if testcase.expectErr {
+			if err == nil {
+				t.Errorf("error expected but not happened for name=%s", testcase.label)
+			} else if err.Error() != testcase.errMsg {
+				t.Errorf("unexpected error message: expected=\"%s\", actual=\"%s\"", testcase.errMsg, err.Error())
+			}
+		} else if !testcase.expectErr && err != nil {
+			t.Errorf("unexpected error for name=%s: %v", testcase.label, err)
+		}
+	}
+}

--- a/main.go
+++ b/main.go
@@ -751,8 +751,7 @@ func (a *app) loadDesiredStateFromYaml(yaml []byte, file string, namespace strin
 	if len(labels) > 0 {
 		err = st.FilterReleases(labels)
 		if err != nil {
-			log.Print(err)
-			return nil, true, nil
+			return nil, false, err
 		}
 	}
 

--- a/state/release_filters.go
+++ b/state/release_filters.go
@@ -62,7 +62,7 @@ func ParseLabels(l string) (LabelFilter, error) {
 			kv := strings.Split(label, "=")
 			lf.positiveLabels = append(lf.positiveLabels, kv)
 		} else { // malformed case
-			err = fmt.Errorf("Malformed label: %s. Expected label in form k=v or k!=v", label)
+			return lf, fmt.Errorf("Malformed label: %s. Expected label in form k=v or k!=v", label)
 		}
 	}
 	return lf, err

